### PR TITLE
[QoL] Filter Bar in Starter Select

### PR DESCRIPTION
**Pull request by FredeX**
### WORK IN PROGRESS

this was often discussed, but the only PR i found working on this was closed due to inactivity. so i opened this draft to keep you updated of my progress. 

**this is a very early stage, so everything is subject to change.**

https://github.com/pagefaultgames/pokerogue/assets/1838682/707095d2-8390-404f-bcf0-b68a55c3e433

UPDATE: I started integrating the gen filter into the filterbar. Therefore i added a scrolling to the container:

https://github.com/pagefaultgames/pokerogue/assets/1838682/e1184503-aae4-4c7e-8a31-2f235454f2cd

### CODE REVIEW

**Comment by HopsWas** [comment]
might be cool to add an option to filter by cost, or perhaps an option to toggle off pokemon you can no longer afford to bring. just some ideas.

**Comment by FredeX** [comment]
> might be cool to add an option to filter by cost, an option to filter by pokemon who you dont have ribbons for, filter by who has enough candy to afford an upgrade, perhaps an option to toggle off pokemon you can no longer afford to bring. just some ideas.

Definitly on the list! First im building the framework for the dropdowns etc, thats why i started with so few filters :) Thanks for your input!

**Comment by FredeX** [comment]
https://github.com/pagefaultgames/pokerogue/assets/1838682/2eb1373d-529f-40e6-bfda-6cef38f08311

**Comment by FredeX** [comment]
Filtering kind of works. But `starter-select-ui-handler.ts` is an absolute hell. Will do some refactoring. 

https://github.com/pagefaultgames/pokerogue/assets/1838682/3c6c8fb4-1cdb-490f-86a2-861d5c17eaab

**Comment by FredeX** [comment]
Okay did most of the refactoring and implemented sorting with ascending and descending orders.

https://github.com/pagefaultgames/pokerogue/assets/1838682/789b1969-594d-48bf-bc2c-3028d7619678

**Comment by FredeX** [comment]
Now is working with Pokerus and the StarterCursor
https://github.com/pagefaultgames/pokerogue/assets/1838682/4a7ec007-ec63-4dd3-894b-7fc9d9ac1e36

**Comment by bennybroseph** [comment]
Now that it looks more feature complete, very interesting way to implement this. Considering this shortens the UI by one row, what are the consequences? How does everything in the later gens still fit? 

Also, I feel like if you're filtering down, you probably don't care about the Gen selector anymore. It'd be nice to just see all pokemon in a scrolling list that fit the criteria at that point and just put another filter for Gen as well. Wondering if others feel the same. Making this it's own window also allows you to customize it more.

**Comment by bennybroseph** [comment]
Actually, yeah what if you just designed the whole Starter UI around this filter. Get rid of the Gen Selection UI, it's not needed if you can filter stuff

**Comment by FredeX** [comment]
> Actually, yeah what if you just designed the whole Starter UI around this filter. Get rid of the Gen Selection UI, it's not needed if you can filter stuff
Yeah i think having two different filters (the gen selector **AND** the filter bar) is kind of unintuitive. Actually this is what i am currently working on.

> Considering this shortens the UI by one row, what are the consequences? How does everything in the later gens still fit?
In the current implementation i actually did **not** cut one row off but rather reduced the spacings to still be able to fit all gens in the container. 
But youre right: when integrating the gen filter im gonna need a scrollable container. i wonder how that feels

**Comment by FredeX** [comment]
Oh and i implemented a few technical gimmics i really enjoy :D 
1. The different filters are automatically spaced evenly across the width of the filter bar. This is because i hate to adjust everything manually especially when i integrate localization, this would get ugly really fast. And maybe ein combination with my "Auto Size Text" PR this will not need any adjustments for the different languages. (The dropdowns are also auto sized, but here the dropdown is the size to fit the biggest option)
2. When a dropdown is too far to the right, its getting fixed to the right
3. 

Screenshot 2024-06-14 070047
Screenshot 2024-06-14 070128

**Comment by FredeX** [comment]
Okay i overhauled the UI a bit and tried out a different layout to incorporate the gen filter into the filterbar.
Do be able to display more than 81 pokemons i changed the container to a custom scrollable container. 

What do you think @bennybroseph ?


https://github.com/pagefaultgames/pokerogue/assets/1838682/bfc3a233-7eb8-4f04-9919-9b1ef8c30b2e

**Comment by flx-sta** [comment]
closes #1770

File: src/ui/starter-select-ui-handler.ts [hunk 1]
```diff
@@ -1709,86 +1679,166 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    this.pokerusCursorObjs.forEach(cursor => cursor.setVisible(false));
+    this.starterCursorObjs.forEach(cursor => cursor.setVisible(false));
 
-      let genCursorWithScroll = this.getGenCursorWithScroll();
+    // filter
+    for (let g=0; g < this.starterContainers.length; g++) {
+      this.starterContainers[g].forEach(container => {
+        container.setVisible(false);
+
+        const speciesVariants = container.species.speciesId && this.scene.gameData.dexData[container.species.speciesId].caughtAttr & DexAttr.SHINY
```
Comment (ReneGV) [inline src/ui/starter-select-ui-handler.ts:1695]: I believe the `& DexAttr.SHINY` makes it so it forces only shiny starters to be showable on the UI

https://github.com/pagefaultgames/pokerogue/assets/31145813/08e2439f-667d-4166-9d17-1d30eb7848ae

Tested locally only Gen 1 shinies showed up, eventhough I had `non shiny` enabled.

File: src/ui/starter-select-ui-handler.ts [hunk 2]
```diff
@@ -301,13 +295,80 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    shiny1.setTint(getVariantTint(2));
+    const shiny2 = this.scene.add.sprite(0, 0, "shiny_star_small");
+    shiny2.setTint(getVariantTint(0));
+    const shiny3 = this.scene.add.sprite(0, 0, "shiny_star_small");
+    shiny3.setTint(getVariantTint(1));
+    const shinyOptions = [
+      new DropDownOption(this.scene, 64n, null, shiny1),
+      new DropDownOption(this.scene, 16n, null, shiny2),
+      new DropDownOption(this.scene, 32n, null, shiny3),
+      new DropDownOption(this.scene, DexAttr.NON_SHINY, "non shiny")];
```
Comment (KimJeongSun) [inline src/ui/starter-select-ui-handler.ts:351]:
```suggestion
    const shiny1Sprite = this.scene.add.sprite(0, 0, "shiny_star_small");
    shiny1Sprite.setTint(getVariantTint(0));
    const shiny2Sprite = this.scene.add.sprite(0, 0, "shiny_star_small");
    shiny2Sprite.setTint(getVariantTint(1));
    const shiny3Sprite = this.scene.add.sprite(0, 0, "shiny_star_small");
    shiny3Sprite.setTint(getVariantTint(2));
    
    const shinyOptions = [
      new DropDownOption(this.scene, "SHINY3", null, shiny3Sprite),
      new DropDownOption(this.scene, "SHINY2", null, shiny2Sprite),
      new DropDownOption(this.scene, "SHINY", null, shiny1Sprite),
      new DropDownOption(this.scene, "NORMAL", "Normal"),
      new DropDownOption(this.scene, "UNCAUGHT", "Not Caught")];
```

File: src/ui/starter-select-ui-handler.ts [hunk 3]
```diff
@@ -1709,86 +1679,166 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
 
-      let genCursorWithScroll = this.getGenCursorWithScroll();
+    // filter
+    for (let g=0; g < this.starterContainers.length; g++) {
+      this.starterContainers[g].forEach(container => {
+        container.setVisible(false);
+
+        const speciesVariants = container.species.speciesId && this.scene.gameData.dexData[container.species.speciesId].caughtAttr & DexAttr.SHINY
+          ? [ DexAttr.DEFAULT_VARIANT, DexAttr.VARIANT_2, DexAttr.VARIANT_3 ].filter(v => !!(this.scene.gameData.dexData[container.species.speciesId].caughtAttr & v))
+          : [];
```
Comment (KimJeongSun) [inline src/ui/starter-select-ui-handler.ts:1697]:
```suggestion
        // First, ensure you have the caught attributes for the species else default to bigint 0
        const caughtVariants = this.scene.gameData.dexData[container.species.speciesId]?.caughtAttr || BigInt(0);
        
        // Define the variables based on whether their respective variants have been caught
        const isVariant3Caught = !!(caughtVariants & DexAttr.VARIANT_3);
        const isVariant2Caught = !!(caughtVariants & DexAttr.VARIANT_2);
        const isVariantCaught = !!(caughtVariants & DexAttr.SHINY);
        const isCaught = !!(caughtVariants & DexAttr.NON_SHINY);
        const isUncaught = !isCaught && !isVariantCaught && !isVariant2Caught && !isVariant3Caught;

```

File: src/ui/starter-select-ui-handler.ts [hunk 4]
```diff
@@ -1709,86 +1679,166 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+        const fitsGen =   this.filterBar.getVals(DropDownColumn.GEN).includes(g);
+        const fitsType =  this.filterBar.getVals(DropDownColumn.TYPES).some(type => container.species.isOfType((type as number) - 1));
+        const fitsShiny = this.filterBar.getVals(DropDownColumn.SHINY).some(variant => speciesVariants.includes(variant));
```
Comment (KimJeongSun) [inline src/ui/starter-select-ui-handler.ts:1700]:
```suggestion
        const fitsShiny = this.filterBar.getVals(DropDownColumn.SHINY).some(variant => {
          if (variant === "SHINY3") {
            return isVariant3Caught;
          } else if (variant === "SHINY2") {
            return isVariant2Caught && !isVariant3Caught;
          } else if (variant === "SHINY") {
            return isVariantCaught && !isVariant2Caught && !isVariant3Caught;
          } else if (variant === "NORMAL") {
            return isCaught && !isVariantCaught && !isVariant2Caught && !isVariant3Caught;
          } else if (variant === "UNCAUGHT") {
            return isUncaught;
          }
        });
```

**Review by KimJeongSun — COMMENTED** [review]
I think this feature is very useful and impressive, so I fixed a bug in the shiny filter part that you mentioned.
For testing, I imported my online data. I can upload it if needed.

Screen Recording 2024-07-04 at 2 42 33 PM (1)

After testing the updated code, it seems that the part you reviewed is now working properly. However, it looks like the original PR author has not been active for 3 weeks. 

If it's okay with you, is there a good way for me to commit my changes so that we can complete this feature?

Additionally, I have an idea: if possible, in the next update of this PR, we could add a function to check whether a passive has been unlocked and include the evolution type in the type filter.

**Comment by KimJeongSun** [comment]
@FredeX 
It seems you don't have the time to finish this PR right now. 
If it's alright with you, how about I create a new PR and complete the filter feature using your commits?
If there's no response by tomorrow, I will proceed with the work in the new PR.
If you have any objections, please let me know and I will try to follow your opinion as the original author.

**Comment by FredeX** [comment]
> @FredeX It seems you don't have the time to finish this PR right now. If it's alright with you, how about I create a new PR and complete the filter feature using your commits? If there's no response by tomorrow, I will proceed with the work in the new PR. If you have any objections, please let me know and I will try to follow your opinion as the original author.
> 
> [related discord message](https://discord.com/channels/1125469663833370665/1176874654015684739/1259736929940013116)

Hey, yeah sadly ive been super busy with personal shit lately and havent found the time to finish up the PR. I would be happy to see it being continued since I wont be able to really work on it for at least 2 more weeks or so.
If you need help with the code or anything hit me up and i try to support where i can.

**Comment by FredeX** [comment]
Im closing this PR for now since Kim opened a follow-up PR. Thank you for continuing the filter. Can't wait to see it in game

**Pull request closed without merging**